control-plane/agents/src/bin/core/volume/registry.rs

```diff
@@ -197,4 +198,72 @@ impl Registry {
             self.notify(event).await;
         }
     }
+
+    /// Get the snapshot states for the specified volume.
+    pub(crate) async fn snap_state(
+        &self,
+        snap_spec: &VolumeSnapshotUserSpec,
+    ) -> Result<VolumeSnapshotState, SvcError> {
+        Ok(VolumeSnapshotState::new(
+            SnapshotInfo::new(snap_spec.source_id(), snap_spec.uuid().clone()),
+            0,
+            None,
+            vec![],
+        ))
+    }
+
+    /// Return a single snapshot object corresponding to the either matching (volumeid + snapid)
+    /// OR matching snapid when input volumeid is None.
+    pub(crate) async fn snapshot(&self, vol_id: Option<&VolumeId>, snap_id: &SnapshotId) -> Result<VolumeSnapshot, SvcError> {
+        let snap_specs = self.specs().snapshots();
+        let spec = snap_specs
+            .iter()
+            .find(|s| 
+                (vol_id.is_some() && s.spec().source_id() == vol_id.unwrap() && s.spec().uuid() == snap_id) || 
+                (!vol_id.is_some() && s.spec().uuid() == snap_id)
+        );
+
+        if spec.is_some() {
+            Ok(VolumeSnapshot::new(
+                VolumeSnapshotDef::from(spec.unwrap()),
+                self.snap_state(spec.unwrap().spec()).await?,
+            ))
+        } else {
+            Err(SvcError::NotFound { kind: ResourceKind::VolumeSnapshot, id: snap_id.to_string() })
+        }
+
+    }
+
+    /// Get all snapshots for a volume.
+    pub(crate) async fn snapshots_all_by_vol(&self, vol_id: &VolumeId) -> Vec<VolumeSnapshot> {
+        let snap_specs = self
+        .specs()
+        .snapshots()
```

Comment:
might be better to add function on specs which does this for us, and void copying the entire list of snapshots.

Reply:
Something like this?

```
pub struct SpecFilterParam<Param: Clone> {
    filter: Param,
}
impl<Param: Clone> SpecFilterParam<Param> {
    pub fn new(filter_by: &Param) -> Self {
        Self { filter: filter_by.clone() }
    }

    /// Get a refernce to the filter parameter set in this object.
    pub fn filter(&self) -> &Param {
        &self.filter
    }
}
```

Then in the specs, create functions that filter and collect specs by `SpecFilterParam<VolumeId>` and `SpecFilterParam<SnapshotId>`